### PR TITLE
Reset desired balance on no longer master

### PR DESCRIPTION
**Pull request by idegtiarenko**
This change resets desired balance (to free memory and prevent it from being reused in future) and completes pending listeners once node is no longer master

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java [hunk 1]
```diff
@@ -155,6 +155,11 @@ public String toString() {
             Priority.URGENT,
             new ReconcileDesiredBalanceExecutor()
         );
+        clusterService.addListener(event -> {
+            if (event.localNodeMaster() == false) {
+                onNoLongerMaster();
```
[the remark this answers is not in the document]
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java:160]: Yeah, probably it is not necessary with this change

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java [hunk 2]
```diff
@@ -281,8 +286,6 @@ public void onFailure(Exception e) {
             if (e.getCause() != null && e.getCause() instanceof EsRejectedExecutionException esRejectedExecutionException) {
                 assert esRejectedExecutionException.isExecutorShutdown();
                 // TODO now what? onNoLongerMaster() asserts it's on the master thread but we could be anywhere here
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java:288]: I reckon we can drop all of this now. Certainly the comment is unnecessary, but nor do we really need to branch here.

**Review by DaveCTurner — APPROVED** [review]
LGTM (one nit)

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java (hunk 2 above)
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java:288]: Sorry, not sure I follow. Do you want to drop `assert esRejectedExecutionException.isExecutorShutdown();` assertion?

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java:288]: Yes, I don't think that assertion is particularly important.

**Pull request merged**